packages/web3-providers-ws/test/integration/ganache_fault_tolerance.test.ts

```diff
@@ -0,0 +1,389 @@
+					mockCallback();
+					resolve(true);
+				});
+			});
+
+			const server = ganache.server();
+			await server.listen(port);
+			await connectPromise;
+			webSocketProvider.disconnect();
+			await server.close();
+			expect(mockCallback).toHaveBeenCalledTimes(1);
+		});
+
+		it('allows disconnection on lost connection, when reconnect is enabled', async () => {
+			const reconnectionOptions = {
+				delay: 10,
+				autoReconnect: true,
+				maxAttempts: 100,
+			};
+			const webSocketProvider = new WebSocketProvider(host, {}, reconnectionOptions);
+
+			const connectPromise = waitForOpenSocketConnection(webSocketProvider);
+
+			const server = ganache.server();
+			await server.listen(port);
+			await connectPromise;
+			webSocketProvider.disconnect();
+			expect(!!(await waitForEvent(webSocketProvider, 'disconnect'))).toBe(true);
+			await server.close();
+		});
+
+		it('errors when failing to reconnect after data is lost mid-chunk', async () => {
+			const server = ganache.server();
+			await server.listen(port);
+			const reconnectionOptions = {
+				delay: 100,
+				autoReconnect: true,
+				maxAttempts: 1,
+			};
+			const mockCallBack = jest.fn();
+			const webSocketProvider = new WebSocketProvider(host, {}, reconnectionOptions);
+			await waitForOpenSocketConnection(webSocketProvider);
+
+			await server.close();
+
+			const errorPromise = new Promise(resolve => {
+				webSocketProvider.on('error', (err: any) => {
+					if (err.message === `Maximum number of reconnect attempts reached! (${1})`) {
+						mockCallBack();
+						resolve(true);
+					}
+				});
+			});
+			// send an event to be parsed and fail
+			const event = {
+				data: 'abc|--|ded',
+				type: 'websocket',
+				// @ts-expect-error run protected method
+				target: webSocketProvider._socketConnection,
+			};
+			// @ts-expect-error run protected method
+			webSocketProvider._onMessage(event);
+			await errorPromise;
+			expect(mockCallBack).toHaveBeenCalled();
+		});
+
+		it('times out when connection is lost mid-chunk', async () => {
+			const server = ganache.server();
+			await server.listen(port);
+			const reconnectionOptions = {
+				delay: 0,
+				autoReconnect: false,
+				maxAttempts: 0,
+			};
+			const webSocketProvider = new WebSocketProvider(host, {}, reconnectionOptions);
+			await waitForOpenSocketConnection(webSocketProvider);
+
+			await server.close();
+
+			const errorPromise = new Promise(resolve => {
+				webSocketProvider.on('error', (err: any) => {
+					if (err.innerError.message === 'Chunk timeout') resolve(true);
+				});
+			});
+			// send an event to be parsed and fail
+			const event = {
+				data: 'abc|--|ded',
+				type: 'websocket',
+				// @ts-expect-error run protected method
+				target: webSocketProvider._socketConnection,
+			};
+			// @ts-expect-error run protected method
+			webSocketProvider._parseResponses(event); // simulate chunks
+			await errorPromise;
+			expect(true).toBe(true);
```

Comment:
I would suggest replacing this with the same `expect(mockCallBack).toHaveBeenCalled();` strategy you've been using for consistency/sanity

Reply:
or maybe replace `if (err.innerError.message === 'Chunk timeout') resolve(true);` with an `expect`:

```typescript
webSocketProvider.on('error', (err: any) => {
    if (err.innerError.message === 'Chunk timeout') resolve(true);
    expect(err).toBeInstanceOf(InvalidResponseError);
    expect(err.innerError.message).toBe('Chunk timeout');
    resolve(true);
});
```

Reply:
you are right thanks, i forgot to create an expect for this. changed